seapath_ros_driver/src/seapath_ros_driver.cpp

```diff
@@ -0,0 +1,285 @@
+    geometry_msgs::msg::TwistWithCovarianceStamped twist_msg;
+    rclcpp::Time current_time;
+    twist_msg.header.stamp = current_time = this->now();
+
+    twist_msg.header.frame_id = "seapath/frame/twist";
+
+    twist_msg.twist.twist.linear.x = data.north_velocity;
+    twist_msg.twist.twist.linear.y = data.east_velocity;
+    twist_msg.twist.twist.linear.z = data.down_velocity;
+
+    twist_msg.twist.twist.angular.x = data.roll_rate;
+    twist_msg.twist.twist.angular.y = data.pitch_rate;
+    twist_msg.twist.twist.angular.z = data.yaw_rate;
+
+    twist_msg.twist.covariance[0] = data.latitude_error * data.latitude_error;
+    twist_msg.twist.covariance[7] = data.longitude_error * data.longitude_error;
+    twist_msg.twist.covariance[14] = data.height_error * data.height_error; 
+
+
+    // Temp hack to avoid inf in covariance
+    if (data.heading_error > 10.0) {
+        twist_msg.twist.covariance[35] = 10.0;
+    } else {
+        twist_msg.twist.covariance[35] = data.heading_error * data.heading_error;
+    }
+    twist_msg.twist.covariance[21] = data.roll_error * data.roll_error; 
+    twist_msg.twist.covariance[28] = data.pitch_error * data.pitch_error;  
+
+    return twist_msg;
+}
+
+
+geometry_msgs::msg::Point SeaPathRosDriver::get_origin_publisher(){
+    
+    geometry_msgs::msg::Point origin_msg;
+    origin_msg.set__x(ORIGIN_N);
+    origin_msg.set__y(ORIGIN_E);
+    origin_msg.set__z(ORIGIN_H);
+
+    RCLCPP_INFO_STREAM(rclcpp::get_logger("rclcpp"), "New Origin: " << "N:" << origin_msg.x << ", E:" << origin_msg.y << "H: " << origin_msg.z <<"\n");
+    return origin_msg;
+}
+
+
+diagnostic_msgs::msg::DiagnosticStatus SeaPathRosDriver::get_diagnostic_publisher(){
+    //check if it's properly connected to the socket
+    diagnostic_msgs::msg::DiagnosticStatus diagnostic_msg;
+    if(seaPathSocket.socket_connected == false){
+        diagnostic_msg.level = diagnostic_msgs::msg::DiagnosticStatus::WARN;
+        diagnostic_msg.name = "Diagnostic_connection_to_socket_status";
+        diagnostic_msg.message = "Socket disconnected";
+    }
+    else if (seaPathSocket.socket_connected == true){
+        diagnostic_msg.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+        diagnostic_msg.name = "Diagnostic_connection_to_socket_status";
+        diagnostic_msg.message = "Socket connection is OK";
+    }
+    return diagnostic_msg;
+}
+
+
+sensor_msgs::msg::NavSatFix SeaPathRosDriver::get_navsatfix_publisher(const KMBinaryData& data){
+    sensor_msgs::msg::NavSatFix nav_msg;
+    rclcpp::Time current_time;
+
+    nav_msg.header.stamp = current_time = this->now();
+    nav_msg.header.frame_id = "seapath/frame/NavSatFix";
+    
+    nav_msg.latitude = data.latitude;
+    nav_msg.longitude = data.longitude;
+    nav_msg.altitude = data.ellipsoid_height;
+
+    return nav_msg;
+}
+vortex_msgs::msg::KMBinaryData kmbinary_publisher(const KMBinaryData& data) {
```

Comment:
This is not a publisher, this is just a function to create the message that is to be published.